File: args.py
import argparse
import os
import sys

import dbs


def get_args():
    # Parse Arguments
    parser = argparse.ArgumentParser(description="yarGen")

    group_creation = parser.add_argument_group("Rule Creation")
    group_creation.add_argument("-m", help="Path to scan for malware")
    group_creation.add_argument(
        "-y",
        help="Minimum string length to consider (default=8)",
        metavar="min-size",
        default=8,
    )
    group_creation.add_argument(
        "-z",
        help="Minimum score to consider (default=0)",
        metavar="min-score",
        default=0,
    )
    group_creation.add_argument(
        "-x",
        help="Score required to set string as 'highly specific string' (default: 30)",
        metavar="high-scoring",
        default=30,
    )
    group_creation.add_argument(
        "-w",
        help="Minimum number of strings that overlap to create a super rule (default: 5)",
        metavar="superrule-overlap",
        default=5,
    )
    group_creation.add_argument(
        "-s",
        help="Maximum length to consider (default=128)",
        metavar="max-size",
        default=128,
        type=int,
    )
    group_creation.add_argument(
        "-rc",
        help="Maximum number of strings per rule (default=20, intelligent filtering "
        "will be applied)",
        metavar="maxstrings",
        default=20,
    )
    group_creation.add_argument(
        "--excludegood",
        help="Force the exclude all goodware strings",
        action="store_true",
        default=False,
    )

    group_output = parser.add_argument_group("Rule Output")
    group_output.add_argument(
        "-o",
        help="Output rule file",
        metavar="output_rule_file",
        default="yargen_rules.yar",
    )
    group_output.add_argument(
        "-e",
        help="Output directory for string exports",
        metavar="output_dir_strings",
        default="",
    )
    group_output.add_argument(
        "-a", help="Author Name", metavar="author", default="yarGen Rule Generator"
    )
    group_output.add_argument(
        "-r",
        help="Reference (can be string or text file)",
        metavar="ref",
        default="https://github.com/Neo23x0/yarGen",
    )
    group_output.add_argument("-l", help="License", metavar="lic", default="")
    group_output.add_argument(
        "-p",
        help="Prefix for the rule description",
        metavar="prefix",
        default="Auto-generated rule",
    )
    group_output.add_argument(
        "-b",
        help="Text file from which the identifier is read (default: last folder name in "
        'the full path, e.g. "myRAT" if -m points to /mnt/mal/myRAT)',
        metavar="identifier",
        default="not set",
    )
    group_output.add_argument(
        "--score",
        help="Show the string scores as comments in the rules",
        action="store_true",
        default=False,
    )
    group_output.add_argument(
        "--strings",
        help="Show the string scores as comments in the rules",
        action="store_true",
        default=False,
    )
    group_output.add_argument(
        "--nosimple",
        help="Skip simple rule creation for files included in super rules",
        action="store_true",
        default=False,
    )
    group_output.add_argument(
        "--nomagic",
        help="Don't include the magic header condition statement",
        action="store_true",
        default=False,
    )
    group_output.add_argument(
        "--nofilesize",
        help="Don't include the filesize condition statement",
        action="store_true",
        default=False,
    )
    group_output.add_argument(
        "-fm",
        help="Multiplier for the maximum 'filesize' condition value (default: 3)",
        default=3,
    )
    group_output.add_argument(
        "--globalrule",
        help="Create global rules (improved rule set speed)",
        action="store_true",
        default=False,
    )
    group_output.add_argument(
        "--nosuper",
        action="store_true",
        default=False,
        help="Don't try to create super rules " "that match against various files",
    )

    group_db = parser.add_argument_group("Database Operations")
    group_db.add_argument(
        "--update",
        action="store_true",
        default=False,
        help="Update the local strings and opcodes " "dbs from the online repository",
    )
    group_db.add_argument(
        "-g",
        help="Path to scan for goodware (dont use the database shipped with yaraGen)",
    )
    group_db.add_argument(
        "-u",
        action="store_true",
        default=False,
        help="Update local standard goodware database with "
        "a new analysis result (used with -g)",
    )
    group_db.add_argument(
        "-c",
        action="store_true",
        default=False,
        help="Create new local goodware database "
        '(use with -g and optionally -i "identifier")',
    )
    group_db.add_argument(
        "-i",
        default="",
        help="Specify an identifier for the newly created databases "
        "(good-strings-identifier.db, good-opcodes-identifier.db)",
    )

    group_general = parser.add_argument_group("General Options")
    group_general.add_argument(
        "--dropzone",
        action="store_true",
        default=False,
        help="Dropzone mode - monitors a directory [-m] for new samples to process. "
        "WARNING: Processed files will be deleted!",
    )
    group_general.add_argument(
        "--nr",
        action="store_true",
        default=False,
        help="Do not recursively scan directories",
    )
    group_general.add_argument(
        "--oe",
        action="store_true",
        default=False,
        help="Only scan executable extensions EXE, "
        "DLL, ASP, JSP, PHP, BIN, INFECTED",
    )
    group_general.add_argument(
        "-fs",
        help="Max file size in MB to analyze (default=10)",
        metavar="size-in-MB",
        default=10,
    )
    group_general.add_argument(
        "--noextras",
        action="store_true",
        default=False,
        help="Don't use extras like Imphash or PE header specifics",
    )
    group_general.add_argument(
        "--ai",
        action="store_true",
        default=False,
        help="Create output to be used as ChatGPT4 input",
    )
    group_general.add_argument(
        "--debug", action="store_true", default=False, help="Debug output"
    )
    group_general.add_argument(
        "--trace", action="store_true", default=False, help="Trace output"
    )

    group_opcode = parser.add_argument_group("Other Features")
    group_opcode.add_argument(
        "--opcodes",
        action="store_true",
        default=False,
        help="Do use the OpCode feature "
        "(use this if not enough high "
        "scoring strings can be found)",
    )
    group_opcode.add_argument(
        "-n",
        help="Number of opcodes to add if not enough high scoring string could be found "
        "(default=3)",
        metavar="opcode-num",
        default=3,
    )

    group_inverse = parser.add_argument_group("Inverse Mode (unstable)")
    group_inverse.add_argument(
        "--inverse", help=argparse.SUPPRESS, action="store_true", default=False
    )
    group_inverse.add_argument(
        "--nodirname", help=argparse.SUPPRESS, action="store_true", default=False
    )
    group_inverse.add_argument(
        "--noscorefilter", help=argparse.SUPPRESS, action="store_true", default=False
    )

    args = parser.parse_args()

    # Print Welcome

    if not args.update and not args.m and not args.g:
        parser.print_help()
        print("")
        print(
            """
[E] You have to select --update to update yarGens database or -m for signature generation or -g for the 
creation of goodware string collections 
(see https://github.com/Neo23x0/yarGen#examples for more details)

Recommended command line:
    python yarGen.py -a 'Your Name' --opcodes --dropzone -m ./dropzone"""
        )
        sys.exit(1)

    # Update
    if args.update:
        dbs.update_databases()
        print("[+] Updated databases - you can now start creating YARA rules")
        sys.exit(0)

    # Typical input erros
    if args.m:
        if os.path.isfile(args.m):
            print("[E] Input is a file, please use a directory instead (-m path)")
            sys.exit(0)

    return args

File: src/args.rs
use clap::{arg, command, ArgGroup, Parser, ValueHint};
use std::path::PathBuf;

#[derive(Parser, Debug, Clone)]
#[command(
    name = "yarobot",
    about = "YARA rule generator",
    version,
    group(
        ArgGroup::new("mode")
            .args(["m", "g", "update"])
            .required(true)
            .multiple(false)
    )
)]
pub struct AppArgs {
    /// Path to scan for malware
    #[arg(short = 'm', value_hint = ValueHint::DirPath)]
    pub m: Option<PathBuf>,

    /// Minimum string length to consider (default=8)
    #[arg(short = 'y', default_value = "8")]
    pub y: usize,

    /// Minimum score to consider (default=5)
    #[arg(short = 'z', default_value = "5")]
    pub z: i32,

    /// Score required to set string as 'highly specific string' (default: 30)
    #[arg(short = 'x', default_value = "30")]
    pub x: i32,

    /// Minimum number of strings that overlap to create a super rule (default: 5)
    #[arg(short = 'w', default_value = "5")]
    pub w: usize,

    /// Maximum length to consider (default=128)
    #[arg(short = 's', default_value = "128")]
    pub s: usize,

    /// Maximum number of strings per rule (default=20, intelligent filtering will be applied)
    #[arg(long = "rc", default_value = "20")]
    pub rc: usize,

    /// Force the exclude all goodware strings
    #[arg(long = "excludegood")]
    pub excludegood: bool,

    /// Output rule file
    #[arg(short = 'o', default_value = "yarobot_rules.yar")]
    pub o: PathBuf,

    /// Output directory for string exports
    #[arg(short = 'e', value_hint = ValueHint::DirPath)]
    pub e: Option<PathBuf>,

    /// Author Name
    #[arg(short = 'a', default_value = "yarobot Rule Generator")]
    pub a: String,

    /// Reference (can be string or text file)
    #[arg(long = "ref", default_value = "https://github.com/oogre2007/yarobot")]
    pub ref_: String,

    /// License
    #[arg(short = 'l')]
    pub l: Option<String>,

    /// Prefix for the rule description
    #[arg(short = 'p', default_value = "Auto-generated rule")]
    pub p: String,

    /// Text file from which the identifier is read
    #[arg(short = 'b', default_value = "not set")]
    pub b: String,

    /// Show the string scores as comments in the rules
    #[arg(long = "score")]
    pub score: bool,

    /// Show the string scores as comments in the rules
    #[arg(long = "strings")]
    pub strings: bool,

    /// Skip simple rule creation for files included in super rules
    #[arg(long = "nosimple")]
    pub nosimple: bool,

    /// Don't include the magic header condition statement
    #[arg(long = "nomagic")]
    pub nomagic: bool,

    /// Don't include the filesize condition statement
    #[arg(long = "nofilesize")]
    pub nofilesize: bool,

    /// Multiplier for the maximum 'filesize' condition value (default: 3)
    #[arg(long = "fm", default_value = "3")]
    pub fm: i32,

    /// Create global rules (improved rule set speed)
    #[arg(long = "globalrule")]
    pub globalrule: bool,

    /// Don't try to create super rules that match against various files
    #[arg(long = "nosuper")]
    pub nosuper: bool,

    /// Update the local strings and opcodes dbs from the online repository
    #[arg(long = "update")]
    pub update: bool,

    /// Path to scan for goodware (dont use the database shipped with yaraGen)
    #[arg(short = 'g', value_hint = ValueHint::DirPath)]
    pub g: Option<PathBuf>,

    /// Update local standard goodware database with a new analysis result (used with -g)
    #[arg(short = 'u')]
    pub u: bool,

    /// Create new local goodware database (use with -g and optionally -i "identifier")
    #[arg(short = 'c')]
    pub c: bool,

    /// Specify an identifier for the newly created databases
    #[arg(short = 'i')]
    pub i: Option<String>,

    /// Dropzone mode - monitors a directory [-m] for new samples to process. WARNING: Processed files will be deleted!
    #[arg(long = "dropzone")]
    pub dropzone: bool,

    /// Recursively scan directories
    #[arg(short = 'R')]
    pub R: bool,

    /// Only scan executable extensions EXE, DLL, ASP, JSP, PHP, BIN, INFECTED
    #[arg(long = "oe")]
    pub oe: bool,

    /// Max file size in MB to analyze (default=10)
    #[arg(long = "fs", default_value = "10")]
    pub fs: usize,

    /// Don't use extras like Imphash or PE header specifics
    #[arg(long = "noextras")]
    pub noextras: bool,

    /// Debug output
    #[arg(long = "debug")]
    pub debug: bool,

    /// Trace output
    #[arg(long = "trace")]
    pub trace: bool,

    /// Do use the OpCode feature (use this if not enough high scoring strings can be found)
    #[arg(long = "opcodes")]
    pub opcodes: bool,

    /// Number of opcodes to add if not enough high scoring string could be found (default=3)
    #[arg(short = 'n', default_value = "3")]
    pub n: i32,
}

impl AppArgs {
    /// Get identifier based on path or file content
    pub fn get_identifier(&self, path: &PathBuf) -> String {
        if self.b == "not set" || !PathBuf::from(&self.b).exists() {
            path.file_name()
                .and_then(|s| s.to_str())
                .unwrap_or("unknown")
                .to_string()
        } else {
            std::fs::read_to_string(&self.b)
                .unwrap_or_else(|_| "unknown".to_string())
                .trim()
                .to_string()
        }
    }

    /// Get reference, reading from file if path exists
    pub fn get_reference(&self) -> String {
        let ref_path = PathBuf::from(&self.ref_);
        if ref_path.exists() {
            std::fs::read_to_string(&ref_path)
                .unwrap_or_else(|_| self.ref_.clone())
                .trim()
                .to_string()
        } else {
            self.ref_.clone()
        }
    }

    /// Get prefix for rule description
    pub fn get_prefix(&self, identifier: &str) -> String {
        if self.p == "Auto-generated rule" {
            identifier.to_string()
        } else {
            self.p.clone()
        }
    }
}
